### DISCUSSION TITLE
Search: Ideas for improvement

### DISCUSSION THREAD

**Discussion by inthreedee** [photoprism/photoprism#1082] [Ideas]
Just some ideas to improve the search capability. 

Some of these things can be achieved using the existing filters, but a smarter search box is way faster than filters. All of our desktop OSes, mobile OSes, and browsers have moved towards smart universal search boxes, so I think searching vs uses filters is behavior everyone is already used to and perhaps even expects. I see this as a long-term incremental improvement project rather than a one-and-done. 

Here are the ideas I have so far, based on things I have actually searched for and wished it found while trying to get my library organized:

1. Search for dates and date ranges
Examples:
	- april 2018
	- april 4 2018
	- april to may 2018
	- 2017 to 2018
	- 2017-2018
	- 2018

2. Search for file names
Examples:
	- IMG_3551
	- 3551

3. Search media types
Examples:
	- video
	- pano
	- panorama
	- video april 2018
	- video from 2018
	- picture april 2018
	- photo april 2018

In cases where there are multiple matches, for example a query matching tags, a date, and a file name, show all of those matches.

**Comment by JamesSwift** [comment]
I like your suggestions. Not essential at the moment, but would be nice to have eventually. I just assumed number 2 (searching by file name) was already how it worked. If not, it would make sense for it to work that way.

I feel the other two suggestions could be simply implemented by a search-and-replace function on the client side before the search is submitted. For example:

- `video` could be searched for and replaced with `video:true`
- For dates a regex or similar could be created to find the various patterns and replace them with `before:2018-00-00 after 2017-00-00`

That way, it would also teach users about the various filter functions as well as being convenient.

Enclosing a search in quotes should make it literal, so `"video"` would do a literal search for `video`. and not replace it with `video:true`.

**Comment by lastzero** [comment]
Keep in mind this may create lots of ambiguities (like in the video example but also with dates or names as there are many) and must work in all languages we support. While it would be nice, it's not an investment we can afford at the moment and a simple proof of concept is of little use. Probably not even GitHub has it, Google yes maybe in some cases. In fact, we already try to recognize file names and do a file name search. Agree with James, that it should be handled mostly client side.

**Comment by inthreedee** [comment]
I understand it’s a bigger project than can be focused on right now, I just thought a place to track ideas would be good.

Handling it client-side could certainly make sense. The point I was trying to make was that UX design is moving towards natural language searches rather than filter boxes.  Not all of my ideas may be doable, but they’re just ideas and a place to start a discussion. 

I’m not sure I see a problem with ambiguities though. If a user searches for a term and it matches a date and a key word and a file name, show all the matches. 

For example, every idea I gave works in Google Photos, but they’re Google and obviously search is their jam. Most of those examples also work in the iOS photos app. I can search for “April 2018” and see the photos from that month. Please don’t underestimate the value of this. It saves an incredible amount of time for the user. I assume the Android photos app works the same way. Flickr most likely does too. It’s pretty standard. 

I’m not sure how GitHub is a relevant example, but if anyone wants other examples of non-photo “smart” search boxes to experiment with to get more ideas, these are some that I’ve used:
- The Windows 10 start menu (Cortana)
- OSX spotlight
- iOS spotlight search
- Gnome Activities search
- KDE Plasma Krunner search

**Comment by graciousgrey** [comment]
I like the idea! But yes this is something that takes time and needs to be well planned. Do you mind if we convert this issue into a discussion?

**Comment by inthreedee** [comment]
Not at all.  It's probably better suited for a discussion format anyway.  I'm sure others have more/better ideas for how to go about improving the search.  Mine are mostly limited to my experience coming from the services I mentioned.

**Comment by inthreedee** [comment]
I've just noticed that to search for a filename, you have to type `name:filename`
I was just trying to search for the filename itself, which was not working.  I honestly had no idea that photoprism has its own search syntax we have to use.  This knowledge alone would have saved me so much time while I was first organizing my library.

I, personally, don't mind having to use a special syntax like that when searching now that I know it exists.  But, in my opinion, it's not very user friendly for non-techies nor is it user-discoverable.  On the surface, it seems like a simple enough thing but normal people just don't think about computers and searches that way.  A "smarter" search engine like is being discussed here would still be a good long-term goal.

On the topic of filename searches specifically, another suggestion to add to the list is partial matches.
Sometimes you don't know exactly what you're looking for, but you know the general vicinity, and that's where partial matches help you find what you want.

For example, assuming the file I want to find is `IMG_2356.MOV`, the following display no results:
- 2356
- 235

**Reply by lastzero** [reply]
They're even documented: https://docs.photoprism.org/user-guide/organize/search/#search-filter-list

We'll get you the best search possible, but need to work on other issues like facial recognition first.

**Reply by inthreedee** [reply]
I understand this is not going to be worked on soon and you have other priorities.  Please don't mistake continued discussion here for added pressure to work on it; that's not the case at all.  I'm just adding further notes and suggestions for when the time eventually comes to work on this.

Thanks for pointing me to the search documentation though!  I read through the documentation before setting up my server and somehow missed that.  It happens, even to the most technically-inclined among us.

**Reply by alexisLefebvre** [reply]
To find `IMG_2356.MOV`, use filter with wildcard: `name:"IMG_23*"` or `name:"*_23*"`, etc.

**Reply by inthreedee** [reply]
@alexislefebvre Wow, thanks for pointing that out!  I never thought to try regex search parameters.  That certainly helps.  I just didn't expect that kind of thing to be needed here.

So it sounds like most of this functionality already exists in the search on the server back-end.  It would just need an interpreter to put the searches of a "normal person" into the proper syntax expected by the server.  That narrows the scope quite a bit!

**Reply by lastzero** [reply]
@inthreedee No worries, it's just my sarcasm. Still a bit amazed about the emphasis on UX for normal people when we've both just "wasted" a lot of time getting hardware video transcoding to work, which seems pretty nerdy :)

**Reply by lastzero** [reply]
`name:"*_23*"` uses regular wildcards like in MS-DOS. Regex is not needed and not supported. You are right, most backend functionality exists. Mainly needs UI polishing & obviously better docs. Help with this is much appreaciated.

**Reply by inthreedee** [reply]
> Still a bit amazed about the emphasis on UX for normal people when we've both just "wasted" a lot of time getting hardware video transcoding to work, which seems pretty nerdy :)

Sure, there's a lot of super nerdy work to get there, but I see so much potential with this project as a self-hosted solution even for non-techies.  Us geeks shouldn't be the only ones who can benefit from escaping the clutches of google ;)  I try to think in terms of "so easy my mother could use it", and then figure out what needs to be done to get there.  I can see a scenario where, eventually, one geek in a family sets this up on a raspberry pi, and then their entire family gets to have their own accounts and use it instead of google photos/flickr.  I'm super excited about the possibilities!

**Reply by alexisLefebvre** [reply]
> @alexislefebvre Wow, thanks for pointing that out! I never thought to try regex search parameters. That certainly helps. I just didn't expect that kind of thing to be needed here.

Actually, that wasn't explained explicitly. Here's a PR to make it clear: https://github.com/photoprism/photoprism-docs/pull/62

**Reply by JamesSwift** [reply]
> Thanks for pointing me to the search documentation though! I read through the documentation before setting up my server and somehow missed that. It happens, even to the most technically-inclined among us.

The exact same thing happened to me. I think the design of having the documentation split into various tabs across the top is easily missed. My eye was drawn the the list down the left and assumed that was all the documentation. Perhaps it could be arranged in a folder/tree-like view?

**Reply by graciousgrey** [reply]
I see, there is room for improvement. I moved the search docs up one level in the navigation for now

**Comment by streiman** [comment]
I have a simpel usecase:

I have the name of persons in keywords ("suchbegriffe') and descriptions. I can search for one like "Jakob" and get the pictures. But when I add a second name like "Andi" I get all the pictures with one of the names (OR). I would like to get only pictures with both persons on them (AND). Is that possible? I couldn't find anything here: https://docs.photoprism.org/user-guide/organize/search/

**Reply by alexisLefebvre** [reply]
I suggest you to open a new discussion. This discussion is about future features, not usage with the current version.

**Reply by streiman** [reply]
You are right. But if it is not possible it should be part of a future search, a simple and, or, not.

**Reply by inthreedee** [reply]
I was curious how Google Photos handles this situation, so I did a test.  The default behavior is when two names are entered, it only shows photos of both people (logical AND).  To perform a Logical OR search, I can type OR in all caps between the two names (lowercase or does not do it).

This makes sense to me as a good, sane default.  I'm sure they have an immense amount of data backing up this decision as the best default for most people.

**Reply by graciousgrey** [reply]
AND search is already on the roadmap :) -->  https://github.com/photoprism/photoprism/issues/881

**Reply by alexisLefebvre** [reply]
@graciousgrey yes that's my issue. :slightly_smiling_face: I didn't shared it because it mentions repeated filters but not an `AND` search with several keywords.

**Reply by graciousgrey** [reply]
True, AND for searches without search filter is an additional use case. 
At least until we support keyword as search filter (https://github.com/photoprism/photoprism/issues/882). One of your tickets as well :D

**Comment by graciousgrey** [comment]
The latest release came with some search improvements. AND search is possible now, see: https://docs.photoprism.org/user-guide/organize/search/

**Comment by inthreedee** [comment]
I still think this kind of improvement would be great, but I think I'd like to see the idea broadened even further.
Can we rename this feature request to "**Natural language search"**?

The reason I bring this up again is because pretty much every time I want to search for a photo, I have to reference the docs to figure out what magical incantation to use to find the specific photo I'm looking for. Will I have to use a specific keyword with a colon after it this time, or will I have to use the drop down filters, or just fall back to going to the calendar month and scrolling aimlessly through a thousand photos until I find the right one. The whole search process feels a bit too programmatic and tedious.

A natural language search would not be as accurate as keyword-triggered or filter-based searches, but it's so much easier for a regular human to remember how to use. Just type vaguely what you're looking for and the database gets queried for any and all of those things, displaying all matching results. For an example of how this works, see Google Photos.

For an initial test implementation, I'd like to see what kind of results get returned if the search queries every filter, date, keyword, color, tag, face, etc down the list of all possible filters. Just show me everything that matches what I've typed in the search box.

**Reply by graciousgrey** [reply]
A next step would be to integrate the existing search filters in the UI so that users can click together their search queries without the need to look up search filters in the docs.

Implementing a good natural language search will take much more time. 
I assume just looking in all kind of fields would probably be slow and as you say not very accurate.

**Reply by inthreedee** [reply]
I hate to always draw comparisons to google photos, but it's what I've switched from so it's hard to avoid. Obviously it's google, search is their jam, and that will be really hard to replicate, but I can type just about anything in the google photos search and it will usually get me to the photo I'm looking for. There may be a dozen other photos that aren't what I'm looking for but, if the one I want is in there on the first page of results, it's done its job more efficiently than clicking through a bunch of filters that slow me down.

But I really dislike using filters, so maybe this is just a me problem. I feel like they're so inefficient and cumbersome to use. Personally, I'll take a less accurate but easier and quicker to use search over having to use a bunch of filters or magic phrases known only to the documentation.
$0.02

**Reply by lastzero** [reply]
Well, I come from professional photo tools and I usually don't know what to enter so that a "smart search" returns what I want.

I don't use Google Photos, but the regular Google search and it works best if I use at least 2 to max 3 related keywords, because that's how their index works. Anything else (using more words, understanding "natural language") can work, but it's like makeup on a pig. It may be Google, but the quality of the results gets bad. So that's that.

Also, we have a lot of Asian users and it would take a long time to understand natural search in Chinese. I guess the best strategy is to make a lot of money so we can hire specialists (for all languages). Otherwise, I think we could provide a neat proof of concept, but that's about it.

**Comment by inthreedee** [comment]
Would it be helpful at all if I started keeping a list of searches I try that fail? 

A few recent examples:
- Tried searching for a date value, didn't recognize it. Used the UI filters to narrow down to year and month. Input the day number in the search field, didn't work. Had to refer to the docs to figure out that I needed to instead use `day:[number]`. (side note, this is so cumbersome a process that I now just search dates in google photos then reference the file name in PhotoPrism)
- Tried searching for `panorama` but had to refer to the docs to figure out I needed to use `panorama:yes`
- Tried searching for `slowmo:yes` but it's not a thing. The docs don't help. A search for `slowmo` returns some but not all of them.
- Discovered that searching for `live` is equivalent to `live:yes`, so this one works fine :tada: 
- Tried searching for `"new york"` (with quotes) but it returns everything except new york. Tried `location:"new york"` and `place:"new york"`, invalid request. Tried `new AND york`, but it didn't find them. Referred to the docs and saw that it has to be `state:"new york"`

I don't really understand why the `keyword:[value]` search format is necessary sometimes but not others, as noted above. I get the impression that the kinds of things I'm searching for should be working and maybe are just oversights? If that is the case, I'm happy to keep listing off those scenarios as I find them so they can be fixed.

I guess, overall, I'd really like to be able to search for either `keyword` (ie panorama) or `value` (ie dates, locations, filenames, etc, with/without quotes) for all the options listed in the docs without using the `keyword:[value]` format.

I can see how this might seem nitpicky but, for someone who hops back and forth between multiple photo tools on various devices and platforms, having to remember PhotoPrism's own unique search "language" is a frustrating experience. So again, I want to try to help improve it and am happy to keep listing off scenarios like the bullet points above. Just let me know if this is helpful.

**Reply by graciousgrey** [reply]
> Would it be helpful at all if I started keeping a list of searches I try that fail?

Sure, that's what discussions are for :) 

Note, however, that the next releases will focus on multi-user features.

Feedback will definitely be valuable once we have resources  to improve search.

**Comment by inthreedee** [comment]
> Anything else (using more words, understanding "natural language") can work, but it's like makeup on a pig. It may be Google, but the quality of the results gets bad. So that's that.

> Also, we have a lot of Asian users and it would take a long time to understand natural search in Chinese. I guess the best strategy is to make a lot of money so we can hire specialists (for all languages). Otherwise, I think we could provide a neat proof of concept, but that's about it.

My how things have changed in just a few short months. Natural language search tech is now available in lots of languages using what OpenAI has developed and it's in the process of being integrated pretty much everywhere. What caught my attention is that DuckDuckGo is rolling out private, natural language search using OpenAI's tech with wikipedia as their source of data in the back end. This suggests that an existing dataset can be fed into the natural language search processor to spit out relevant and accurate responses.

Sounds like, eventually, Photoprism will be able to benefit from this as well. As long as the AI doesn't become sentient and nuke us all, right?

Reference, DDG's blog: https://spreadprivacy.com/duckassist-launch/

**Reply by lastzero** [reply]
You're right! Once we have a little time, we'll certainly look into it. Improved sharing and batch editing seems to be more important though.

**Reply by inthreedee** [reply]
I came across a list of open alternatives to chatgpt. It's probably worth waiting for things to settle and develop a bit more before putting any time into this sort of thing, but it's exciting to see the progress that's happening out there:
https://github.com/nichtdax/awesome-totally-open-chatgpt

**Comment by poisonnuke** [comment]
As a temporary improvement is there a way to set AND as default search option when searching multiple keywords? Its extremely unnatural (and useless too) to have OR as default.
